engine/backends/coq/coq/coq_backend.ml

```diff
@@ -88,598 +104,2219 @@ module Context = struct
   type t = { current_namespace : string * string list }
 end
 
-let primitive_to_string (id : primitive_ident) : string =
-  match id with
-  | Deref -> "(TODO: Deref)" (* failwith "Deref" *)
-  | Cast -> "cast" (* failwith "Cast" *)
-  | LogicalOp op -> ( match op with And -> "andb" | Or -> "orb")
+let hardcoded_coq_headers =
+  "(* File automatically generated by Hacspec *)\n\
+   From Coq Require Import ZArith.\n\
+   Require Import List.\n\
+   Import List.ListNotations.\n\
+   Open Scope Z_scope.\n\
+   Open Scope bool_scope.\n\
+   Require Import Ascii.\n\
+   Require Import String.\n\
+   Require Import Coq.Floats.Floats.\n\
+   From RecordUpdate Require Import RecordSet.\n\
+   Import RecordSetNotations.\n"
 
-module Make (Ctx : sig
-  val ctx : Context.t
-end) =
+module Make (Default : sig
+  val default : string -> string
+end)
+(Attrs : Attrs.WITH_ITEMS) =
 struct
-  open Ctx
-
-  let pconcrete_ident (id : concrete_ident) : string =
-    let id = U.Concrete_ident_view.to_view id in
-    let crate, path = ctx.current_namespace in
-    if String.(crate = id.crate) && [%eq: string list] id.path path then
-      id.definition
-    else
-      (* id.crate ^ "_" ^ *)
-      (* List.fold_left ~init:"" ~f:(fun x y -> x ^ "_" ^ y) *)
-      id.definition
-
-  let pglobal_ident (id : global_ident) : string =
-    match id with
-    | `Projector (`Concrete cid) | `Concrete cid -> pconcrete_ident cid
-    | `Primitive p_id -> primitive_to_string p_id
-    | `TupleType _i -> "TODO (global ident) tuple type"
-    | `TupleCons _i -> "TODO (global ident) tuple cons"
-    | `Projector (`TupleField _) | `TupleField _ ->
-        "TODO (global ident) tuple field"
-    | _ -> .
-
-  module TODOs_debug = struct
-    let __TODO_pat__ _ s = C.AST.Ident (s ^ " todo(pat)")
-    let __TODO_ty__ _ s : C.AST.ty = C.AST.NameTy (s ^ " todo(ty)")
-    let __TODO_item__ _ s = C.AST.Unimplemented (s ^ " todo(item)")
-    let __TODO_term__ _ s = C.AST.Const (C.AST.Const_string (s ^ " todo(term)"))
-  end
+  module Base = Generic_printer.Make (InputLanguage)
+  open PPrint
 
-  module TODOs = struct
-    let __TODO_ty__ span s : C.AST.ty =
-      Error.unimplemented ~details:("[ty] node " ^ s) span
+  let default_string_for s = "TODO: please implement the method `" ^ s ^ "`"
+  let default_document_for = default_string_for >> string
 
-    let __TODO_pat__ span s =
-      Error.unimplemented ~details:("[pat] node " ^ s) span
+  let is_document_empty doc : bool =
+    (* Replace the following with: [PPrint.is_empty], however this currently does not work, bug? *)
+    let buf = Buffer.create 0 in
+    PPrint.ToBuffer.pretty 1.0 80 buf doc;
+    String.equal "" (Buffer.contents buf)
```

Comment:
As we discussed the other day, that will break stuff.
Rendering a document alters the state of the printer, we should *never* do that manually.

Whenever we render, we also write in some internal state some span information; this will break it.